File: ddd-core/src/main/kotlin/com/akuchars/goals/core/domain/AbstractJpaEntity.java
package com.akuchars.goals.core.domain;

import java.util.Objects;

import javax.persistence.Access;
import javax.persistence.AccessType;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.MappedSuperclass;

import org.jetbrains.annotations.Nullable;

@MappedSuperclass
@Access(AccessType.FIELD)
public abstract class AbstractJpaEntity {
	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	@Nullable
	protected Long id;

	public AbstractJpaEntity() {
	}

	@Nullable
	public Long getId() {
		return id;
	}

	public void setId(@Nullable Long id) {
		this.id = id;
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (o == null || getClass() != o.getClass()) return false;
		AbstractJpaEntity that = (AbstractJpaEntity) o;
		return Objects.equals(id, that.id);
	}

	@Override
	public int hashCode() {
		return Objects.hash(id);
	}
}

File: ddd-core/src/main/kotlin/com/akuchars/goals/core/domain/AbstractJpaEntity.kt
package com.akuchars.goals.core.domain

import javax.persistence.GeneratedValue
import javax.persistence.Id
import javax.persistence.MappedSuperclass

@MappedSuperclass
abstract class AbstractJpaEntity(@Id
                              @GeneratedValue
                              var id: Long? = null) {

    override fun toString() = "Entity of type ${this.javaClass.name} with id: $id"

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as AbstractJpaEntity
        if (id != other.id)  return false

        return true
    }

    override fun hashCode(): Int = id?.hashCode() ?: 0

    companion object {
        private val serialVersionUID = -5554308939380869754L
    }

}